File: scripts/preprocessing/gee/javascript/functions/gap_filling.js
/**
 * Applies Inverse Distance Weighting (IDW) interpolation to fill gaps in the image.
 * @param {ee.Image} image - The input image with gaps to fill.
 * @param {ee.Geometry} aoi - The area of interest for interpolation.
 * @param {number} range - The maximum distance (in meters) to search for values.
 * @param {number} gamma - The decay factor for the inverse distance.
 * @param {number} numPixels - The number of pixels to sample for interpolation.
 * @returns {ee.Image} - The image with gaps filled by interpolation.
 */
function applyIDWInterpolation(image, aoi, range, gamma, numPixels) {
  // Get the list of band names as an ee.List
  var bandNames = image.bandNames();

  // Function to interpolate a single band
  var interpolateBand = function(bandName) {
    bandName = ee.String(bandName);
    // Sample the image band to get known values
    var samples = image.select([bandName]).addBands(ee.Image.pixelLonLat())
      .sample({
        region: aoi,
        numPixels: numPixels,
        scale: 30,
        projection: 'EPSG:4326'
      })
      .map(function(sample) {
        var lat = sample.get('latitude');
        var lon = sample.get('longitude');
        var value = sample.get(bandName);
        // Return a feature with the band name set dynamically
        return ee.Feature(ee.Geometry.Point([lon, lat])).set(bandName, value);
      });

    // Estimate global mean and standard deviation from the samples
    var stats = samples.reduceColumns({
      reducer: ee.Reducer.mean().combine({
        reducer2: ee.Reducer.stdDev(),
        sharedInputs: true,
      }),
      selectors: [bandName]
    });

    // Apply IDW interpolation
    var interpolated = samples.inverseDistance({
      range: range,
      propertyName: bandName,
      mean: stats.get('mean'),
      stdDev: stats.get('stdDev'),
      gamma: gamma
    });

    // Return the interpolated band
    return interpolated.rename(bandName);
  };

  // Map the interpolation function over all band names
  var interpolatedBands = bandNames.map(function(bandName) {
    return interpolateBand(bandName);
  });

  // Combine all interpolated bands into a single image
  var interpolatedImage = ee.ImageCollection(interpolatedBands).toBands().clip(aoi);

  return interpolatedImage;
}

exports.applyIDWInterpolation = applyIDWInterpolation;

File: scripts/preprocessing/gee/python/utils/gap_filling.py
# ---
# title:   Gap Filling Functions
# author:  Brendan Casey
# created: 2026-07-10
# notes:
#   Fill gaps in Earth Engine imagery using Inverse
#   Distance Weighting (IDW) interpolation, applied band by
#   band. Each band is sampled to points, interpolated, and
#   recombined into a single multi-band image.
# ---

import ee


def apply_idw_interpolation(image, aoi, range_m, gamma,
                            num_pixels):
    """Fill image gaps with IDW interpolation.

    Parameters
    ----------
    image : ee.Image
        Input image with gaps to fill.
    aoi : ee.Geometry
        Area of interest for interpolation.
    range_m : float
        Maximum distance (in meters) to search for values.
    gamma : float
        Decay factor for the inverse distance weighting.
    num_pixels : int
        Number of pixels to sample for interpolation.

    Returns
    -------
    ee.Image
        The image with gaps filled by interpolation.
    """
    band_names = image.bandNames()

    # Interpolate a single band by name.
    def interpolate_band(band_name):
        band_name = ee.String(band_name)

        # Turn each sampled pixel into a point feature.
        def to_point(sample):
            lat = sample.get("latitude")
            lon = sample.get("longitude")
            value = sample.get(band_name)
            return ee.Feature(
                ee.Geometry.Point([lon, lat])
            ).set(band_name, value)

        # Sample the band to get known values.
        samples = (
            image.select([band_name])
            .addBands(ee.Image.pixelLonLat())
            .sample(
                region=aoi,
                numPixels=num_pixels,
                scale=30,
                projection="EPSG:4326",
            )
            .map(to_point)
        )

        # Global mean and standard deviation of samples.
        stats = samples.reduceColumns(
            reducer=ee.Reducer.mean().combine(
                reducer2=ee.Reducer.stdDev(),
                sharedInputs=True,
            ),
            selectors=[band_name],
        )

        # Apply IDW interpolation.
        interpolated = samples.inverseDistance(
            range=range_m,
            propertyName=band_name,
            mean=stats.get("mean"),
            stdDev=stats.get("stdDev"),
            gamma=gamma,
        )

        return interpolated.rename(band_name)

    # Interpolate every band and combine into one image.
    interpolated_bands = band_names.map(interpolate_band)

    interpolated_image = (
        ee.ImageCollection(interpolated_bands)
        .toBands()
        .clip(aoi)
    )

    return interpolated_image
